nixos/modules/services/security/yubikey-agent.nix

```diff
@@ -0,0 +1,88 @@
+# Global configuration for yubikey-agent.
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.yubikey-agent;
+
+  # Copied from the gnupg module
+  xserverCfg = config.services.xserver;
+  defaultPinentryFlavor =
+    if xserverCfg.desktopManager.lxqt.enable
+    || xserverCfg.desktopManager.plasma5.enable then
+      "qt"
+    else if xserverCfg.desktopManager.xfce.enable then
+      "gtk2"
+    else if xserverCfg.enable || config.programs.sway.enable then
+      "gnome3"
+    else
+      null;
+
+in
+{
+  ###### interface
+
+  meta.maintainers = [ maintainers.philandstuff ];
+
+  options = {
+
+    services.yubikey-agent = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to start yubikey-agent when you log in.  Also sets
+          SSH_AUTH_SOCK to point at yubikey-agent.
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.yubikey-agent;
+        defaultText = "pkgs.yubikey-agent";
+        description = ''
+          The package used for the yubikey-agent daemon.
+        '';
+      };
+
+      pinentryFlavor = mkOption {
+        type = types.nullOr (types.enum pkgs.pinentry.flavors);
+        example = "gnome3";
+        description = ''
+        Which pinentry interface to use. If not null, the path to the
+        pinentry binary will be provided to yubikey-agent by setting the PATH
+        appropriately in the systemd service and thus overrides the pinentry
+        option in the yubikey-agent package's default systemd unit.
+        If not set at all, it'll pick an appropriate flavor depending on the
+        system configuration (qt flavor for lxqt and plasma5, gtk2 for xfce
+        4.12, gnome3 on all other systems with X enabled, ncurses otherwise).
+      '';
+      };
+
+    };
+  };
+
+  config = {
+    services.yubikey-agent.pinentryFlavor = mkDefault defaultPinentryFlavor;
+
+    environment.systemPackages = [ cfg.package ];
+    systemd.packages = [ cfg.package ];
+
+    # This overrides the systemd user unit shipped with the
+    # yubikey-agent package
+    systemd.user.services.yubikey-agent = mkIf (cfg.pinentryFlavor != null) {
```

Comment:
I tried doing this, but I got:

```
error: The option `programs.gnupg.agent.pinentryFlavor' is used but not defined.
(use '--show-trace' to show detailed location information)
```

I'm not sure what I'm doing wrong here, or if there's something clever I need to do to depend on another module's configuration.

Reply:
If works if you don't assign the whole `config` attrset at once:

```diff
index d378666b4f1..4c07b9227d6 100644
--- a/nixos/modules/services/security/yubikey-agent.nix
+++ b/nixos/modules/services/security/yubikey-agent.nix
@@ -7,19 +7,6 @@ with lib;
 let
   cfg = config.services.yubikey-agent;
 
-  # Copied from the gnupg module
-  xserverCfg = config.services.xserver;
-  defaultPinentryFlavor =
-    if xserverCfg.desktopManager.lxqt.enable
-    || xserverCfg.desktopManager.plasma5.enable then
-      "qt"
-    else if xserverCfg.desktopManager.xfce.enable then
-      "gtk2"
-    else if xserverCfg.enable || config.programs.sway.enable then
-      "gnome3"
-    else
-      null;
-
 in
 {
   ###### interface
@@ -46,41 +33,26 @@ in
           The package used for the yubikey-agent daemon.
         '';
       };
-
-      pinentryFlavor = mkOption {
-        type = types.nullOr (types.enum pkgs.pinentry.flavors);
-        example = "gnome3";
-        description = ''
-        Which pinentry interface to use. If not null, the path to the
-        pinentry binary will be provided to yubikey-agent by setting the PATH
-        appropriately in the systemd service and thus overrides the pinentry
-        option in the yubikey-agent package's default systemd unit.
-        If not set at all, it'll pick an appropriate flavor depending on the
-        system configuration (qt flavor for lxqt and plasma5, gtk2 for xfce
-        4.12, gnome3 on all other systems with X enabled, ncurses otherwise).
-      '';
-      };
-
     };
   };
 
-  config = {
-    services.yubikey-agent.pinentryFlavor = mkDefault defaultPinentryFlavor;
 
-    environment.systemPackages = [ cfg.package ];
-    systemd.packages = [ cfg.package ];
+  config.environment = {
+    extraInit = optionalString cfg.enable ''
+      if [ -z "$SSH_AUTH_SOCK" -a -n "$XDG_RUNTIME_DIR" ]; then
+        export SSH_AUTH_SOCK="$XDG_RUNTIME_DIR/yubikey-agent/yubikey-agent.sock"
+      fi
+    '';
+    systemPackages = [ cfg.package ];
+  };
+
+  config.systemd = {
+    packages = [ cfg.package ];
 
     # This overrides the systemd user unit shipped with the
     # yubikey-agent package
-    systemd.user.services.yubikey-agent = mkIf (cfg.pinentryFlavor != null) {
-      path = [ pkgs.pinentry.${cfg.pinentryFlavor} ];
+    user.services.yubikey-agent = mkIf (config.programs.gnupg.agent.pinentryFlavor != null) {
+      path = [ pkgs.pinentry.${config.programs.gnupg.agent.pinentryFlavor} ];
     };
-
-    environment.extraInit = optionalString cfg.enable
-      ''
-        if [ -z "$SSH_AUTH_SOCK" -a -n "$XDG_RUNTIME_DIR" ]; then
-          export SSH_AUTH_SOCK="$XDG_RUNTIME_DIR/yubikey-agent/yubikey-agent.sock"
-        fi
-      '';
   };
 }
```

Reply:
I applied this patch, exactly as in your message, on top of d883cfad, but I still got the same problem.

I've found the issue: my configuation.nix doesn't set `programs.gnupg.agent.enable` or `programs.gnupg.agent.pinentryFlavor`.  The gnupg mkDefault is only evaluated if `programs.gnupg.agent.enable` is true; which means you can only get the useful defaulting behaviour if you choose to run the gpg-agent.

It would be possible, but not easy, to edit the gnupg module to always apply the `mkDefault`, but I'm not sure it's the right solution for us here.

I don't like the current copypaste solution but I don't think we've found anything better quite yet. Maybe we could extract a separate `pinentry` module, which just does the defaulting behaviour, and reference it from `gnupg` and `yubikey-agent`?